File: app/src/main/java/com/manimarank/spell4wiki/activities/ContributorsActivity.java
package com.manimarank.spell4wiki.activities;

import android.os.Bundle;
import android.os.Handler;
import android.view.MenuItem;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.interpolator.view.animation.FastOutSlowInInterpolator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.tabs.TabLayout;
import com.manimarank.spell4wiki.R;
import com.manimarank.spell4wiki.adapters.ContributorsAdapter;
import com.manimarank.spell4wiki.adapters.CoreContributorsAdapter;
import com.manimarank.spell4wiki.apis.ApiClient;
import com.manimarank.spell4wiki.apis.ApiInterface;
import com.manimarank.spell4wiki.models.ContributorData;
import com.manimarank.spell4wiki.models.Contributors;
import com.manimarank.spell4wiki.models.CoreContributors;
import com.manimarank.spell4wiki.utils.NetworkUtils;
import com.manimarank.spell4wiki.utils.ShowCasePref;
import com.manimarank.spell4wiki.utils.SnackBarUtils;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import uk.co.samuelwall.materialtaptargetprompt.MaterialTapTargetPrompt;
import uk.co.samuelwall.materialtaptargetprompt.MaterialTapTargetSequence;
import uk.co.samuelwall.materialtaptargetprompt.extras.focals.RectanglePromptFocal;


public class ContributorsActivity extends AppCompatActivity {

    private List<Contributors> contributorsList = new ArrayList<>();
    private List<CoreContributors> coreContributorsList = new ArrayList<>();
    private RecyclerView recyclerViewCodeContributors, recyclerViewCoreContributors;
    private AppCompatTextView txtHelpers;
    private View loadingContributors, layoutCoreContributors;
    private CoreContributorsAdapter coreContributorsAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_contributors);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setTitle(getString(R.string.contributors));
        }

        recyclerViewCoreContributors = findViewById(R.id.recyclerViewCoreContributors);
        recyclerViewCodeContributors = findViewById(R.id.recyclerViewCodeContributors);
        txtHelpers = findViewById(R.id.txtHelpers);
        loadingContributors = findViewById(R.id.loadingContributors);
        layoutCoreContributors = findViewById(R.id.layoutCoreContributors);

        ContributorsAdapter contributorsAdapter = new ContributorsAdapter(this, contributorsList);
        recyclerViewCodeContributors.setAdapter(contributorsAdapter);
        recyclerViewCodeContributors.setLayoutManager(new LinearLayoutManager(getApplicationContext()));

        coreContributorsAdapter = new CoreContributorsAdapter(this, coreContributorsList);
        recyclerViewCoreContributors.setAdapter(coreContributorsAdapter);
        recyclerViewCoreContributors.setLayoutManager(new LinearLayoutManager(getApplicationContext()));

        loadCoreContributorsAndHelpersFromApi();
        TabLayout tabLayout = findViewById(R.id.tabLayout);
        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                if (tab.getPosition() == 1)
                    loadCodeContributorsFromApi();
                else
                    loadCoreContributorsAndHelpersFromApi();
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });
    }

    private void loadCoreContributorsAndHelpersFromApi() {

        loadingContributors.setVisibility(View.VISIBLE);
        layoutCoreContributors.setVisibility(View.GONE);
        recyclerViewCodeContributors.setVisibility(View.GONE);

        if (NetworkUtils.INSTANCE.isConnected(getApplicationContext())) {

            ApiInterface api = ApiClient.getApi().create(ApiInterface.class);
            Call<ContributorData> call = api.fetchContributorData();

            call.enqueue(new Callback<ContributorData>() {
                @Override
                public void onResponse(@NotNull Call<ContributorData> call, @NotNull Response<ContributorData> response) {
                    if (response.isSuccessful() && response.body() != null) {
                        coreContributorsList.clear();
                        coreContributorsList.addAll(response.body().getCore_contributors());
                        if (coreContributorsAdapter != null)
                            coreContributorsAdapter.notifyDataSetChanged();

                        StringBuilder wikiTechHelpers = new StringBuilder();
                        response.body().getWiki_tech_helpers();
                        if (response.body().getWiki_tech_helpers().size() > 0) {
                            for (String helper : response.body().getWiki_tech_helpers()) {
                                wikiTechHelpers.append(helper).append("\n");
                            }
                            txtHelpers.setText(wikiTechHelpers.toString());
                        }

                        if (loadingContributors != null)
                            loadingContributors.setVisibility(View.GONE);
                        if (layoutCoreContributors != null)
                            layoutCoreContributors.setVisibility(View.VISIBLE);

                        new Handler().postDelayed(() -> callShowCaseUI(), 1000);
                    }

                }

                @Override
                public void onFailure(@NotNull Call<ContributorData> call, @NotNull Throwable t) {
                    t.printStackTrace();
                }
            });
        } else {
            SnackBarUtils.INSTANCE.showLong(recyclerViewCodeContributors, getString(R.string.check_internet));
        }
    }

    private void loadCodeContributorsFromApi() {
        loadingContributors.setVisibility(View.VISIBLE);
        recyclerViewCodeContributors.setVisibility(View.GONE);
        layoutCoreContributors.setVisibility(View.GONE);

        ApiInterface api = ApiClient.getApi().create(ApiInterface.class);
        Call<List<Contributors>> call = api.fetchCodeContributorsList();

        call.enqueue(new Callback<List<Contributors>>() {
            @Override
            public void onResponse(@NotNull Call<List<Contributors>> call, @NotNull Response<List<Contributors>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    contributorsList.clear();
                    contributorsList.addAll(response.body());
                    if (recyclerViewCodeContributors != null && recyclerViewCodeContributors.getAdapter() != null)
                        recyclerViewCodeContributors.getAdapter().notifyDataSetChanged();
                }

                if (loadingContributors != null)
                    loadingContributors.setVisibility(View.GONE);
                if (recyclerViewCodeContributors != null)
                    recyclerViewCodeContributors.setVisibility(View.VISIBLE);
            }

            @Override
            public void onFailure(@NotNull Call<List<Contributors>> call, @NotNull Throwable t) {
                t.printStackTrace();
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        if (menuItem.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return (super.onOptionsItemSelected(menuItem));
    }

    private void callShowCaseUI() {
        if (!isFinishing() && !isDestroyed() && ShowCasePref.INSTANCE.isNotShowed(ShowCasePref.CORE_CONTRIBUTORS_LIST_ITEM) && recyclerViewCoreContributors != null && recyclerViewCoreContributors.getVisibility() == View.VISIBLE && recyclerViewCoreContributors.getChildAt(0) != null) {
            MaterialTapTargetSequence sequence = new MaterialTapTargetSequence().setSequenceCompleteListener(() -> ShowCasePref.INSTANCE.showed(ShowCasePref.CORE_CONTRIBUTORS_LIST_ITEM));
            sequence.addPrompt(new MaterialTapTargetPrompt.Builder(ContributorsActivity.this)
                    .setPromptFocal(new RectanglePromptFocal())
                    .setAnimationInterpolator(new FastOutSlowInInterpolator())
                    .setFocalPadding(R.dimen.show_case_focal_padding)
                    .setTarget(recyclerViewCoreContributors.getChildAt(0))
                    .setPrimaryText(R.string.sc_t_core_contributors_list_item)
                    .setSecondaryText(R.string.sc_d_core_contributors_list_item));
            sequence.show();
        }
    }

}

File: app/src/main/java/com/manimarank/spell4wiki/ui/about/ContributorsActivity.kt
package com.manimarank.spell4wiki.ui.about

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.MenuItem
import android.view.View
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.manimarank.spell4wiki.R
import com.manimarank.spell4wiki.data.apis.ApiClient
import com.manimarank.spell4wiki.data.apis.ApiInterface
import com.manimarank.spell4wiki.data.model.CodeContributors
import com.manimarank.spell4wiki.data.model.ContributorData
import com.manimarank.spell4wiki.data.model.CoreContributors
import com.manimarank.spell4wiki.data.prefs.ShowCasePref
import com.manimarank.spell4wiki.data.prefs.ShowCasePref.isNotShowed
import com.manimarank.spell4wiki.data.prefs.ShowCasePref.showed
import com.manimarank.spell4wiki.ui.common.BaseActivity
import com.manimarank.spell4wiki.utils.EdgeToEdgeUtils.setupEdgeToEdgeWithToolbar
import com.manimarank.spell4wiki.utils.GeneralUtils
import com.manimarank.spell4wiki.utils.NetworkUtils.isConnected
import com.manimarank.spell4wiki.utils.SnackBarUtils.showLong
import com.manimarank.spell4wiki.utils.makeGone
import com.manimarank.spell4wiki.utils.makeVisible
import com.manimarank.spell4wiki.databinding.ActivityContributorsBinding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import uk.co.samuelwall.materialtaptargetprompt.MaterialTapTargetSequence
import java.util.*

class ContributorsActivity : BaseActivity() {

    private lateinit var binding: ActivityContributorsBinding
    private val codeContributorsList: MutableList<CodeContributors> = ArrayList()
    private val coreContributorsList: MutableList<CoreContributors> = ArrayList()
    private lateinit var coreContributorsAdapter: CoreContributorsAdapter
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityContributorsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Setup toolbar
        val toolbar = findViewById<androidx.appcompat.widget.Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        // Setup edge-to-edge display
        setupEdgeToEdgeWithToolbar(
            rootView = binding.root,
            toolbar = toolbar
        )

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = getString(R.string.contributors)

        val contributorsAdapter = ContributorsAdapter(this, codeContributorsList)
        binding.recyclerViewCodeContributors.adapter = contributorsAdapter
        binding.recyclerViewCodeContributors.layoutManager = GridLayoutManager(applicationContext, 2)
        coreContributorsAdapter = CoreContributorsAdapter(this, coreContributorsList)
        binding.recyclerViewCoreContributors.adapter = coreContributorsAdapter
        binding.recyclerViewCoreContributors.layoutManager = LinearLayoutManager(applicationContext)

        // Setup tabs programmatically
        setupTabs()

        loadCoreContributorsAndHelpersFromApi()
    }

    private fun setupTabs() {
        val tabLayout = binding.tabLayout

        // Add tabs programmatically
        tabLayout.addTab(tabLayout.newTab().setText(getString(R.string.core_contributors)))
        tabLayout.addTab(tabLayout.newTab().setText(getString(R.string.code_contributors)))

        // Set up tab selection listener
        tabLayout.addOnTabSelectedListener(object : OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                when (tab.position) {
                    0 -> loadCoreContributorsAndHelpersFromApi()
                    1 -> loadCodeContributorsFromApi()
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {}
            override fun onTabReselected(tab: TabLayout.Tab) {}
        })
    }

    private fun loadCoreContributorsAndHelpersFromApi() {
        binding.loadingContributors.root.makeVisible()
        binding.layoutCoreContributors.makeGone()
        binding.recyclerViewCodeContributors.makeGone()
        if (isConnected(applicationContext)) {
            val api = ApiClient.api.create(ApiInterface::class.java)
            val call = api.fetchContributorData()
            call.enqueue(object : Callback<ContributorData?> {
                override fun onResponse(call: Call<ContributorData?>, response: Response<ContributorData?>) {
                    if (response.isSuccessful && response.body() != null) {
                        val resBody = response.body()
                        coreContributorsList.clear()
                        resBody?.core_contributors?.forEach { coreContributors ->
                            coreContributorsList.add(coreContributors)
                        }
                        coreContributorsAdapter.notifyDataSetChanged()
                        val wikiTechHelpers = StringBuilder()
                        resBody?.wiki_tech_helpers?.forEach { helper ->
                            wikiTechHelpers.append("- ").append(helper).append("\n")
                        }
                        binding.txtHelpers.text = wikiTechHelpers.toString()
                        binding.loadingContributors.root.makeGone()
                        binding.layoutCoreContributors.makeVisible()
                        Handler(Looper.getMainLooper()).postDelayed({ callShowCaseUI() }, 1000)
                    }
                }

                override fun onFailure(call: Call<ContributorData?>, t: Throwable) {
                    t.printStackTrace()
                    binding.loadingContributors.root.makeGone()
                    showLong(binding.recyclerViewCodeContributors, getString(R.string.something_went_wrong))
                }
            })
        } else {
            binding.loadingContributors.root.makeGone()
            showLong(binding.recyclerViewCodeContributors, getString(R.string.check_internet))
        }
    }

    private fun loadCodeContributorsFromApi() {
        binding.loadingContributors.root.makeVisible()
        binding.recyclerViewCodeContributors.makeGone()
        binding.layoutCoreContributors.makeGone()

        if (isConnected(applicationContext)) {
            val api = ApiClient.api.create(ApiInterface::class.java)
            val call = api.fetchCodeContributorsList()
            call.enqueue(object : Callback<List<CodeContributors?>?> {
                override fun onResponse(call: Call<List<CodeContributors?>?>, response: Response<List<CodeContributors?>?>) {
                    if (response.isSuccessful && response.body() != null) {
                        codeContributorsList.clear()
                        response.body()?.filterNotNull()?.forEach { cc ->
                            codeContributorsList.add(cc)
                        }
                        binding.recyclerViewCodeContributors.adapter?.notifyDataSetChanged()
                    }
                    binding.loadingContributors.root.makeGone()
                    binding.recyclerViewCodeContributors.makeVisible()
                }

                override fun onFailure(call: Call<List<CodeContributors?>?>, t: Throwable) {
                    t.printStackTrace()
                    binding.loadingContributors.root.makeGone()
                    showLong(binding.recyclerViewCodeContributors, getString(R.string.something_went_wrong))
                }
            })
        } else {
            binding.loadingContributors.root.makeGone()
            showLong(binding.recyclerViewCodeContributors, getString(R.string.check_internet))
        }
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        if (menuItem.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(menuItem)
    }

    private fun callShowCaseUI() {
        if (!isFinishing && !isDestroyed && isNotShowed(ShowCasePref.CORE_CONTRIBUTORS_LIST_ITEM) && binding.recyclerViewCoreContributors.visibility == View.VISIBLE && binding.recyclerViewCoreContributors.getChildAt(0) != null) {
            val sequence = MaterialTapTargetSequence().setSequenceCompleteListener { showed(ShowCasePref.CORE_CONTRIBUTORS_LIST_ITEM) }
            sequence.addPrompt(GeneralUtils.getPromptBuilder(this)
                    .setTarget(binding.recyclerViewCoreContributors.getChildAt(0))
                    .setPrimaryText(R.string.sc_t_core_contributors_list_item)
                    .setSecondaryText(R.string.sc_d_core_contributors_list_item))
            sequence.show()
        }
    }
}